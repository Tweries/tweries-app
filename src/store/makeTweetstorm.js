import v from 'voca';

const LINEFEED = '[..]';
const MAX_LENGTH = 280;
const PREFIX_PLACEHOLDER = '_/_';

function makeSequenceNumber(index, length) {
  if (index === undefined && length === undefined) {
    return PREFIX_PLACEHOLDER;
  }
  return `${index + 1}/${length}`;
}

function makeTweetstorm(source, hashtags) {
  let copy = source.slice();
  const parts = [];

  while (copy.length !== 0) {
    let take = v.prune(
      copy,
      MAX_LENGTH - makeSequenceNumber().length - hashtags.length - 2, // INFO: 1 space after the prefix and one space before the suffix
      ''
    );
    if (take.indexOf(LINEFEED) > -1) {
      take = v.substr(take, 0, take.indexOf(LINEFEED));
      copy = v.substr(copy, take.length + 4); // INFO: 4 is the lenght of the linefeed
    } else {
      copy = v.substr(copy, take.length + 1); // INFO: 1 is the space after the word
    }
    parts.push(take);
  }

  const tweetstorm = parts.map((part, index) => {
    const tweet = `${v.trim(part)} ${hashtags} ${makeSequenceNumber(
      index,
      parts.length
    )}`;
    return { length: tweet.length, tweet };
  });

  return tweetstorm;
}

export default makeTweetstorm;
