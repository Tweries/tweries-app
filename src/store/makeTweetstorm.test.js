import makeTweetstorm from './makeTweetstorm';

describe('source => tweetstorm', () => {
  const scenarios = [
    {
      description: 'Lorem ipsum',
      props: {
        hashtags: '#agile #prodmgmt',
        source:
          'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nec tincidunt praesent semper feugiat nibh sed pulvinar. Duis at tellus at urna condimentum mattis pellentesque id nibh. Quisque egestas diam in arcu cursus euismod quis viverra. Mollis nunc sed id semper risus. Aliquet eget sit amet tellus cras adipiscing enim. Pellentesque dignissim enim sit amet venenatis urna cursus eget. Tortor consequat id porta nibh venenatis cras. Sagittis nisl rhoncus mattis rhoncus urna neque viverra justo nec. Egestas egestas fringilla phasellus faucibus. Nulla facilisi etiam dignissim diam quis enim. Maecenas ultricies mi eget mauris pharetra. Ac turpis egestas sed tempus urna et pharetra pharetra. Sagittis nisl rhoncus mattis rhoncus. Purus in mollis nunc sed id semper risus in. Nisl tincidunt eget nullam non nisi est sit amet. Condimentum vitae sapien pellentesque habitant morbi tristique. Integer quis auctor elit sed vulputate mi sit amet mauris. Orci dapibus ultrices in iaculis nunc. Rhoncus est pellentesque elit ullamcorper. Ullamcorper malesuada proin libero nunc consequat interdum varius. Sit amet nisl suscipit adipiscing bibendum est ultricies integer quis.'
      }
    },
    {
      description: 'About the Windsor knot',
      props: {
        hashtags: '#howto #tieatie',
        source:
          'Although the Duke of Windsor never specifically used the Windsor knot, he did favor a wide triangular knot. In actuality, the Duke achieved his trendsetting look by tying a Four-in-Hand with specially made wide and extra thick ties. The Windsor knot was invented by the public as a way to imitate the Duke\'s knot style. There are several derivatives of the Windsor that are all referred to by the same name. The Windsor delivers a symmetrical and solid triangular knot that works best with a spread collar. This knot is also mistakenly referred to as the "Double Windsor" knot.'
      }
    },
    {
      description: 'w/ linefeeds',
      props: {
        hashtags: '#QueenFTW',
        source: `Mama, just killed a man
        Put a gun against his head
        Pulled my trigger, now he's dead
        Mama, life had just begun
        But now I've gone and thrown it all away
        [..]
        Mama, ooh
        Didn't mean to make you cry
        If I'm not back again this time tomorrow
        Carry on, carry on
        As if nothing really matters
        [..]
        Too late, my time has come
        Sends shivers down my spine
        Body's aching all the time
        Goodbye everybody, I've got to go
        Gotta leave you all behind and face the truth`
      }
    },
    {
      description: 'No final comma in the 1st tweet',
      props: {
        hashtags: '',
        source:
          '123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, WHAT?'
      }
    },
    {
      description: 'pick-your-own-linefeed',
      props: {
        hashtags: '',
        linefeed: ';;',
        source: 'first ;; second ;; third ;; fourth'
      }
    }
  ];

  scenarios.forEach(({ description, props }) => {
    test(description, () => {
      const tweetstorm = makeTweetstorm(props);
      expect(tweetstorm).toMatchSnapshot();
    });
  });
});
