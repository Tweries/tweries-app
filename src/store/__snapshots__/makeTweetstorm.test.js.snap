// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source => tweetstorm About the Windsor knot 1`] = `
Array [
  Object {
    "length": 252,
    "tweet": "Although the Duke of Windsor never specifically used the Windsor knot, he did favor a wide triangular knot. In actuality, the Duke achieved his trendsetting look by tying a Four-in-Hand with specially made wide and extra thick ties. #howto #tieatie 1/3",
  },
  Object {
    "length": 194,
    "tweet": "The Windsor knot was invented by the public as a way to imitate the Duke's knot style. There are several derivatives of the Windsor that are all referred to by the same name. #howto #tieatie 2/3",
  },
  Object {
    "length": 189,
    "tweet": "The Windsor delivers a symmetrical and solid triangular knot that works best with a spread collar. This knot is also mistakenly referred to as the \\"Double Windsor\\" knot. #howto #tieatie 3/3",
  },
]
`;

exports[`source => tweetstorm Lorem ipsum 1`] = `
Array [
  Object {
    "length": 265,
    "tweet": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nec tincidunt praesent semper feugiat nibh sed pulvinar. Duis at tellus at urna condimentum mattis pellentesque id nibh. #agile #prodmgmt 1/6",
  },
  Object {
    "length": 227,
    "tweet": "Quisque egestas diam in arcu cursus euismod quis viverra. Mollis nunc sed id semper risus. Aliquet eget sit amet tellus cras adipiscing enim. Pellentesque dignissim enim sit amet venenatis urna cursus eget. #agile #prodmgmt 2/6",
  },
  Object {
    "length": 271,
    "tweet": "Tortor consequat id porta nibh venenatis cras. Sagittis nisl rhoncus mattis rhoncus urna neque viverra justo nec. Egestas egestas fringilla phasellus faucibus. Nulla facilisi etiam dignissim diam quis enim. Maecenas ultricies mi eget mauris pharetra. #agile #prodmgmt 3/6",
  },
  Object {
    "length": 209,
    "tweet": "Ac turpis egestas sed tempus urna et pharetra pharetra. Sagittis nisl rhoncus mattis rhoncus. Purus in mollis nunc sed id semper risus in. Nisl tincidunt eget nullam non nisi est sit amet. #agile #prodmgmt 4/6",
  },
  Object {
    "length": 225,
    "tweet": "Condimentum vitae sapien pellentesque habitant morbi tristique. Integer quis auctor elit sed vulputate mi sit amet mauris. Orci dapibus ultrices in iaculis nunc. Rhoncus est pellentesque elit ullamcorper. #agile #prodmgmt 5/6",
  },
  Object {
    "length": 158,
    "tweet": "Ullamcorper malesuada proin libero nunc consequat interdum varius. Sit amet nisl suscipit adipiscing bibendum est ultricies integer quis. #agile #prodmgmt 6/6",
  },
]
`;

exports[`source => tweetstorm No final comma in the 1st tweet 1`] = `
Array [
  Object {
    "length": 267,
    "tweet": "123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 123456789, 1/2",
  },
  Object {
    "length": 20,
    "tweet": "123456789, WHAT? 2/2",
  },
]
`;

exports[`source => tweetstorm empty linefeed 1`] = `
Array [
  Object {
    "length": 243,
    "tweet": "It put a huge amount of stress on me. Resentment built up and I delegated less and less. With so much on my plate, I was running the risk of burnout, and as my team saw this spiral, they became demotivated and the vicious spiral continued. 1/1",
  },
]
`;

exports[`source => tweetstorm new lines 1`] = `
Array [
  Object {
    "length": 151,
    "tweet": "In many real world uses of the javascript code it's probably better to replace the line breaks with a space instead of just removing them entirely. 1/3",
  },
  Object {
    "length": 191,
    "tweet": "If for instance we got a column of text from a pdf and we just removed the line breaks entirely we might wind up with words and sentences that run together and are hard to read like this: 2/3",
  },
  Object {
    "length": 93,
    "tweet": "Here is a sentence.This sentence runs into itand unfortunately we aremissing some spaces. 3/3",
  },
]
`;

exports[`source => tweetstorm pick-your-own-linefeed 1`] = `
Array [
  Object {
    "length": 9,
    "tweet": "first 1/4",
  },
  Object {
    "length": 10,
    "tweet": "second 2/4",
  },
  Object {
    "length": 9,
    "tweet": "third 3/4",
  },
  Object {
    "length": 10,
    "tweet": "fourth 4/4",
  },
]
`;

exports[`source => tweetstorm text without line breaks 1`] = `
Array [
  Object {
    "length": 250,
    "tweet": "Nel mezzo del cammin di nostra vita mi ritrovai per una selva oscura, ché la diritta via era smarrita. Ahi quanto a dir qual era è cosa dura esta selva selvaggia e aspra e forte che nel pensier rinova la paura! Tant’ è amara che poco è più morte; 1/5",
  },
  Object {
    "length": 233,
    "tweet": "ma per trattar del ben ch’i’ vi trovai, dirò de l’altre cose ch’i’ v’ho scorte. Io non so ben ridir com’ i’ v’intrai,10 tant’ era pien di sonno a quel punto che la verace via abbandonai. Ma poi ch’i’ fui al piè d’un colle giunto, 2/5",
  },
  Object {
    "length": 214,
    "tweet": "là dove terminava quella valle che m’avea di paura il cor compunto, guardai in alto e vidi le sue spalle vestite già de’ raggi del pianeta che mena dritto altrui per ogne calle. Allor fu la paura un poco queta, 3/5",
  },
  Object {
    "length": 223,
    "tweet": "che nel lago del cor m’era durata20 la notte ch’i’ passai con tanta pieta. E come quei che con lena affannata, uscito fuor del pelago a la riva, si volge a l’acqua perigliosa e guata, così l’animo mio, ch’ancor fuggiva, 4/5",
  },
  Object {
    "length": 76,
    "tweet": "si volse a retro a rimirar lo passo che non lasciò già mai persona viva. 5/5",
  },
]
`;

exports[`source => tweetstorm w/ linefeeds 1`] = `
Array [
  Object {
    "length": 196,
    "tweet": "Mama, just killed a man
        Put a gun against his head
        Pulled my trigger, now he's dead
        Mama, life had just begun
        But now I've gone and thrown it all away #QueenFTW 1/4",
  },
  Object {
    "length": 172,
    "tweet": "Mama, ooh
        Didn't mean to make you cry
        If I'm not back again this time tomorrow
        Carry on, carry on
        As if nothing really matters #QueenFTW 2/4",
  },
  Object {
    "length": 138,
    "tweet": "Too late, my time has come
        Sends shivers down my spine
        Body's aching all the time
        Goodbye everybody, #QueenFTW 3/4",
  },
  Object {
    "length": 82,
    "tweet": "I've got to go
        Gotta leave you all behind and face the truth #QueenFTW 4/4",
  },
]
`;
